### DISCUSSION TITLE
Setting up a glib build using meson on unsupported system (gnu/kfreebsd)

### DISCUSSION THREAD

**Discussion by matthegg5** [mesonbuild/meson#11625] [Q&A]
Hello

Firstly, apologies if I'm writing this question in the wrong place, or I should be asking the glib folks.

I'm currently building a bit of a frankenstein system using a Debian-kFreeBSD base then upgrading and compiling everything else from scratch.

I'm trying to set up and compile glib  (obtained via 'git clone https://gitlab.gnome.org/GNOME/glib.git)  using the below command:-

meson setup _build

but it is failing with the below

ERROR: Problem encountered: Unsupported pair: system "gnu/kfreebsd", cpu family "x86_64"

Is this something specific to meson or glib? If the former, is there a way I can change the "system" to think it is Linux, or straight FreeBSD?

Thank you for any assistance

**Comment by eli-schwartz** [comment]
It's difficult to tell without the actual error message (it includes a filename and line number at the beginning) and preferably the full log or at least surrounding context.

If I had to take a wild guess, including via a code search of every repository on GitHub, the answer is probably that libffi's meson-based wrap doesn't support kfreebsd. Do you have a separately installed version of it?

**Reply by matthegg5** [reply]
Wild guess not far off! Log attached.

I don't have a separately-installed libffi as far as I know.

[meson-log.txt](https://github.com/mesonbuild/meson/files/11106117/meson-log.txt)

**Reply by xclaesse** [reply]
It comes from libffi subproject, not sure if glib uses latest meson port from https://github.com/libffi/libffi/pull/746.

**Reply by eli-schwartz** [reply]
I would suggest building a standalone libffi (it will also tend to be faster when you need to rebuild glib, to not build libffi too...)

**Reply by xclaesse** [reply]
The libffi PR linked above is supported to be on par with autotools, would be great to fix the support for your platform. Your Debian base doesn't have libffi already?

**Reply by matthegg5** [reply]
Thanks, no it doesn't have libffi - it's not mainstream Debian, working off an unstable branch of a port, and pretty off-piste from the port now anyway.

I've just built my own libffi from scratch and tried to rerun 'meson setup _build' for glib but it still fails on the libffi subproject error as presented in the earlier logfile. Do I need to change something for meson to not hit this error?

**Reply by eli-schwartz** [reply]
You need to be able to run `pkg-config --cflags --libs "libffi >=3.0.0 "` and have that succeed and print some flags for using libffi.

Double-checking your log, all dependency lookups fail because you do not seem to have pkg-config installed.

**Reply by xclaesse** [reply]
I'm currious how we can fix the Meson port thought. It seems `host_machine.system()` returns `gnu/kfreebsd` which should not be a valid value for libffi autotools neither: https://github.com/libffi/libffi/blob/master/configure.host.

**Reply by matthegg5** [reply]
Thanks

when running the pkg-config command as suggested, I get some flags and paths. Re-running the meson setup _build for glib no longer presents the libffi issue about my wonky combination. Interestingly, I saw that gnu/kfreebsd on x86 (32-bit) is valid according to libffi's github page, but that's beside the matter.

Now hitting a No getxattr implementation found in C library or libxattr issue, but that's for me to explore. I believe we have resolved the issue raised by building and installing an independent libffi. For anyone who sees this in the future, pkg-config depends on glib but glib depends on pkg-config, however you can `./configure --use-internal-glib` on the pkg-config source install to build it.

 Thanks again for your assistance.

**Reply by xclaesse** [reply]
> pkg-config depends on glib but glib depends on pkg-config

Indeed, you could also use pkgconf instead which does not have that issue.
